dartagnan/src/main/java/com/dat3m/dartagnan/encoding/NonTerminationEncoder.java

```diff
@@ -0,0 +1,768 @@
+               and they have to interact "nicely" with each other.
+
+    Approach:
+        (1) All loop iterations of non-terminating loops are decomposed into prefix, infix, and suffix.
+            - The last (non-terminating) iteration of the loop is always in the suffix (but preceding iterations may
+              also be part of the suffix).
+            - For every iteration in the infix, there must be a matching (~equivalent) iteration in the suffix.
+              NOTE: The infix can actually be empty.
+            - We employ a special rule:
+               (i) a side-effect-free non-termination (spinloop) will have empty infix and exactly
+                   a single suffix iteration.
+              (ii) a side-effectful non-termination will have at least one infix iteration
+                   (this means there must be at least two unrollings)
+            --------------
+            - Events belong to infix/suffix IFF any iteration they are contained in belongs to infix/suffix
+              (notice that for nested loops, a single event can be part of multiple iterations/loops).
+            - All other events belong to prefix, in particular, all events outside of any loops.
+        (2) Infix and suffix must be equivalent:
+            - Events in infix must have an equivalent event in suffix with same value/address
+              (TODO: Can maybe be restricted to equivalence on non-dead events)
+            - rf/co-edges between infix events must have an equivalent edge in the suffix.
+            - NOTE: In the special case where we have no infix but a suffix, the only requirement
+              is that the local state at the beginning and the end of the suffix match.
+              This is always true for side-effect-free loops, the only one's where we do not have an infix!
+        (3) Suffix must be strongly(*) fair/consistent:
+            - Prefix stores must be co-before suffix stores
+            - Only suffix reads can read from suffix stores, prefix/infix reads can not.
+            - Suffix reads can only read from prefix if there are no same-address stores in infix/suffix.
+              In that case, it further has to read the co-maximal store.
+              Actually, this requirement is just that there are no fr-edges from suffix to prefix,
+              but since we might not have encoded fr (lazy solving), we formulate this indirectly.
+
+              (*) "Strongness" here is a property that guarantees that not only the finite execution we see is
+                  consistent but that it will stay consistent when we repeatedly append the suffix.
+                  In particular, the infinite repetition of the suffix remains consistent (and fair).
+
+    ASSUMPTIONS:
+      - We assume the NonterminationDetection and DynamicSpinLoopDetection passes have marked
+        possible points of non-termination (otherwise the detection simply fails).
```

Comment:
"detection simply fails" -> which detection are you referring to? The one done by any of the two mentioned passes, some detection done in this class, or the encoding done by this class?

Reply:
How can it refer to any of the two passes if the assumption is that they have already run? :)
It refers to this class and both remaining interpretations are correct: the class does not encode non-termination detection for non-instrumented code and therefore doesn't report issues in those loops.

I updated the comment slightly. Is it any better now?